src/repr/src/adt/numeric.rs

```diff
@@ -762,11 +763,161 @@ impl DecimalLike for Numeric {
     }
 }
 
+/// An encoded packed variant of [`Numeric`].
+///
+/// We uphold the invariant that [`PackedNumeric`] sorts the same as
+/// [`Numeric`].
+///
+/// # Encoding Format
+///
+/// The encoded format of [`PackedNumeric`] is derived from the internal
+/// representation of [`Numeric`], whose details can be found at [1] or [2].
+/// The exact format of [`PackedNumeric`] though is as follows:
+///
+/// `[ marker<4> | exponent<4> | digits<0-26> | num_digits<4> ]`
+///
+/// total encoded size: 38 bytes, allocated size: 40 bytes
+///
+/// * `marker` represents special states of the number, i.e. `-Infinity`,
+/// `-X`, `X`, `Infinity`, or `NaN`.
+/// * `digits` are the unadjusted numeric digits of the number, e.g. for
+///   `123.45` digits would contain `12345`.
+/// * `exponent` represents where to place the decimal in `digits`, in the
+///    previous example it would be `-2`.
+/// * `num_digits` represents the number of digits for the number, e.g. for
+///    `100000000` it would be `9`.
+///
+/// ## Sorting:
+///
+/// We encode everything as big endian and adjust negative values so the byte
+/// representation sorts the same as the numeric representation.
+///
+/// * **marker** contains 5 possible values:
+///     * -Infinity: `192` `hex: 0xC0`
+///     * -X       : `128` `hex: 0x80`
+///     * X        : `96`  `hex: 0x60`
+///     * Infinity : `64`  `hex: 0x40`
+///     * NaN      : `32`  `hex: 0x20`
+///   * These values are exactly what `libdecimal` uses internally, except for
+///     `X`, where the marker value is `0`. We map `X` to `96` so the marker
+///     has the same sort order as [`OrderedDecimal`].
+///   * We invert the `marker` value so the sort order is as follows, from
+///     least to greatest: `-Infinity`, `-X`, `X`, `Infinity`, `NaN`.
+/// * **exponent** is an `i32` that is interpreted as a `u32` and XOR'd with
+///   `0x8000_0000` so negative values will sort before positive ones.
+/// * **digits** are encoded as big endian, with the most significant values
+///   first, and inverted if the number is negative.
+/// * **num_digits** is encoded as big endian.
+///
+/// ## Notes:
+///
+/// * Even though the encoded format only requires 38 bytes, we allocate 40 for
+///   better cache alignment, since 40 is a multiple of 8.
+///
+/// [1]: <https://github.com/gcc-mirror/gcc/blob/2b438a0d2aa80f051a09b245a58f643540d4004b/libdecnumber/decNumber.h#L76-L102>
+/// [2]: <http://speleotrove.com/decimal/dnnumb.html>
+/// [`OrderedDecimal`]: dec::OrderedDecimal
```

Comment:
Thanks - this is extremely helpful!

Reply:
No problem! Just a heads up though, I found a subtle bug in this impl and decided to ditch it 😅

Reply:
😅